porcupine/plugins/git_status.py

```diff
@@ -0,0 +1,250 @@
+                    continue
+            else:
+                # Handle directories containing files with different statuses
+                substatuses = {
+                    s
+                    for p, s in path_to_status.items()
+                    if s in {"git_added", "git_modified", "git_mergeconflict"}
+                    and item_path in p.parents
+                }
+
+                if "git_mergeconflict" in substatuses:
+                    status = "git_mergeconflict"
+                elif "git_modified" in substatuses:
+                    status = "git_modified"
+                elif "git_added" in substatuses:
+                    status = "git_added"
+                else:
+                    assert not substatuses
+                    status = None
+
+            # TODO: allow other plugins to create tags too
+            # Ideally without making two tcl calls.
+            # If I remember correctly this code is a bit perf critical
+            self.tree.item(item_id, tags=([] if status is None else status))
+            if item_id in selection:
+                update_tree_selection_color(self.tree)
+
+    def update_colors_now_or_later(self, item_id: str) -> None:
+        self.coloring_queue.append(item_id)
+        assert self.git_status_future is not None
+        if self.git_status_future.done():
+            self._handle_queue()
+
+
+# not project-specific
+class TreeColorer:
+    def __init__(self, tree: DirectoryTree):
+        self.tree = tree
+        self.project_specific_colorers: dict[str, ProjectColorer] = {}
+
+    def config_color_tags(self, junk: object = None) -> None:
+        fg = self.tree.tk.eval("ttk::style lookup Treeview -foreground")
+        bg = self.tree.tk.eval("ttk::style lookup Treeview -background")
+        gray = utils.mix_colors(fg, bg, 0.5)
+
+        if sum(self.tree.winfo_rgb(fg)) > 3 * 0x7FFF:
+            # bright foreground color
+            green = "#00ff00"
+            orange = "#ff6e00"
+        else:
+            green = "#007f00"
+            orange = "#e66300"
+
+        self.tree.tag_configure("git_mergeconflict", foreground=orange)
+        self.tree.tag_configure("git_modified", foreground="red")
+        self.tree.tag_configure("git_added", foreground=green)
+        self.tree.tag_configure("git_untracked", foreground="red4")
+        self.tree.tag_configure("git_ignored", foreground=gray)
+
+    def start_status_coloring_for_all_projects(self, junk_event: object) -> None:
+        old_project_ids = set(self.project_specific_colorers.keys())
+        new_project_ids = set(self.tree.get_children())
+        for removed in old_project_ids - new_project_ids:
+            del self.project_specific_colorers[removed]
+        for added in new_project_ids - old_project_ids:
+            self.project_specific_colorers[added] = ProjectColorer(self.tree, added)
+
+        for colorer in self.project_specific_colorers.values():
+            colorer.start_running_git_status()
```

Comment:
I'm not sure if this makes sense. It can make old ProjectColorers to start running `git status` even if it's already running. I think a better solution would be to stop all old coloring and start anew.